src/services/postgres_rfqt_firm_quote_validator.ts

```diff
@@ -0,0 +1,188 @@
+    private readonly _cacheExpiryThresholdMs: number;
+
+    private static _calculateTakerFillableAmountsFromQuotes(
+        quotes: SignedOrder[],
+        makerLookup: { [key: string]: BigNumber },
+    ): { makerAddressesToAddToCache: string[]; takerFillableAmounts: BigNumber[] } {
+        const makerAddressesToAddToCacheSet: Set<string> = new Set();
+        const takerFillableAmounts = quotes.map(quote => {
+            const makerTokenBalanceForMaker: BigNumber | undefined = makerLookup[quote.makerAddress];
+
+            // TODO: Add Prometheus hooks
+            if (makerTokenBalanceForMaker === undefined) {
+                makerAddressesToAddToCacheSet.add(quote.makerAddress);
+                return quote.takerAssetAmount;
+            }
+
+            // Order is fully fillable, because Maker has 100% of the assets
+            if (makerTokenBalanceForMaker.gte(quote.makerAssetAmount)) {
+                return quote.takerAssetAmount;
+            }
+
+            // Order is empty, return zero
+            if (quote.makerAssetAmount.lte(0)) {
+                return ZERO;
+            }
+
+            // Order is partially fillable, because Maker has a fraction of the assets
+            const partialFillableAmount = makerTokenBalanceForMaker
+                .times(quote.takerAssetAmount)
+                .div(quote.makerAssetAmount)
+                .integerValue(BigNumber.ROUND_DOWN);
+            logger.warn(
+                `Maker ${
+                    quote.makerAddress
+                } balance is ${makerTokenBalanceForMaker.toString()} and can only partially cover order size ${quote.makerAssetAmount.toString()}. takerAssetAmount was reduced from ${quote.takerAssetAmount.toString()} to ${partialFillableAmount.toString()}`,
+            );
+            if (!partialFillableAmount.isFinite()) {
+                logger.error(
+                    `Calculated maker token balance is infinite, which caused the partialFillableAmount to be infinite. This should never happen`,
+                );
+                return ZERO;
+            }
+            return partialFillableAmount;
+        });
+        const makerAddressesToAddToCache = Array.from(makerAddressesToAddToCacheSet);
+        return {
+            makerAddressesToAddToCache,
+            takerFillableAmounts,
+        };
+    }
+
+    constructor(
+        chainCacheRepository: Repository<MakerBalanceChainCacheEntity>,
+        cacheExpiryThresholdMs: number = THRESHOLD_CACHE_EXPIRED_MS,
+    ) {
+        this._chainCacheRepository = chainCacheRepository;
+        this._cacheExpiryThresholdMs = cacheExpiryThresholdMs;
+    }
+
+    // tslint:disable-next-line: prefer-function-over-method
+    public async getRfqtTakerFillableAmountsAsync(quotes: SignedOrder[]): Promise<BigNumber[]> {
+        // TODO: Handle error on query
+
+        // Ensure that all quotes have the same exact maker token.
```

Comment:
Is it necessary to enforce this? Maybe we'll use RFQT in multi-hop at some point and will request different tokens